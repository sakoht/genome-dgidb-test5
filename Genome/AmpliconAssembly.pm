package Genome::AmpliconAssembly;

use strict;
use warnings;

use Genome;

use Carp 'confess';
use Data::Dumper 'Dumper';
require File::Copy;
require Genome::Consed::Directory;
require Genome::AmpliconAssembly::Amplicon;

my %ATTRIBUTES = (
    directory => {
        is => 'Text',
        doc => 'Base directory',
    },
    sequencing_center => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_sequencing_center,
        doc => 'Sequencing Center that the amplicons were sequenced.  Currently supported centers: '.join(', ', __PACKAGE__->valid_sequencing_centers).' (default: '.__PACKAGE__->default_sequencing_center.').',
    },
    sequencing_platform => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_sequencing_platform,
        doc => 'Platform upon whence the amplicons were sequenced.  Currently supported platforms: '.join(', ', __PACKAGE__->valid_sequencing_platforms).' (default: '.__PACKAGE__->default_sequencing_platform.').',
    },
    assembler => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_assembler,
        doc => 'The assembler to use.  Currently supported assemblers: '.join(', ', valid_assemblers()).' (default: '.__PACKAGE__->default_assembler.').',
    },
    assembly_size => { # amplicon_size
        is => 'Integer',
        is_optional => 1,
        doc => 'Expected assembly size for an assembled amplicon.',
    },
    subject_name => {
        is => 'Text',
        is_optional => 1,
        doc => 'The subject name of the underlying data.  Used primarily in naming files (etc) that represent the entire amplicon assembly, like an assembled fasta of each amplicon.',
    },
);

#< UR >#
class Genome::AmpliconAssembly {
    is => 'UR::Object',
    has => [ %ATTRIBUTES ],
};

sub create {
    my $class = shift;

    my $self = $class->SUPER::create(@_)
        or return;

    unless ( Genome::Utility::FileSystem->validate_existing_directory( $self->directory ) ) {
        $self->delete;
        return;
    }

    for my $attribute (qw/ sequencing_platform sequencing_center assembler /) {
        unless ( $self->validate_attribute_value($attribute) ) {
            $self->delete;
            return;
        }
    }

    $self->create_directory_structure
        or return;

    return $self;
}

#< Atributes >#
sub attributes {
    return %ATTRIBUTES;
}

sub attribute_names {
    return keys %ATTRIBUTES;
}

sub get_attribute_for_name {
    my ($class, $name) = @_;

    confess "No attribute name given to get attribute" unless $name;
    confess "No attribute found for name ($name)" unless exists $ATTRIBUTES{$name};

    return $ATTRIBUTES{$name};
}

sub helpful_methods {
    return (qw/ 
        chromat_dir phd_dir edit_dir
        consed_directory create_directory_structure
        get_amplicons 
        amplicon_fasta_types amplicon_bioseq_method_for_type
        fasta_file_for_type qual_file_for_type
        assembly_fasta reads_fasta processed_fasta 
        /);
}

sub validate_attribute_value {
    my ($self, $attribute) = @_;

    my $value = $self->$attribute;
    unless ( defined $value ) { # all attrs have defaults, make sure it is set
        $self->error_message("No value for $attribute given");
        return;
    }
    
    my $valid_values_method = 'valid_'.$attribute.'s';
    unless ( grep { $value eq $_ } $self->$valid_values_method ) {
        $self->error_message(
            sprintf(
                'Invalid %s %s. Valid %ss: %s',
                $attribute,
                $value,
                $attribute,
                join(', ',valid_sequencing_centers()),
            )
        );
        return;
    }

    return 1;
}

# Sequencing Centers
sub valid_sequencing_centers {
    return (qw/ gsc broad /);
}

sub default_sequencing_center {
    return (valid_sequencing_centers)[0];
}

# Sequencing Platforms
sub valid_sequencing_platforms {
    return (qw/ sanger /);
}

sub default_sequencing_platform {
    return (valid_sequencing_platforms)[0];
}

# Assemblers
sub valid_assemblers {
    return (qw/ phredphrap /);
}

sub default_assembler {
    return (valid_assemblers)[0];
}

#< DIRS >#
sub consed_directory {
    my $self = shift;

    unless ( $self->{_consed_directory} ) {
        $self->{_consed_directory} = Genome::Consed::Directory->create(directory => $self->directory);
    }

    return $self->{_consed_directory};
}

sub create_directory_structure {
    my $self = shift;

    $self->consed_directory->create_extended_directory_structure
        or return;

    return 1;
}

sub edit_dir {
    return $_[0]->consed_directory->edit_dir;
}
    
sub phd_dir {
    return $_[0]->consed_directory->phd_dir;
}
    
sub chromat_dir {
    return $_[0]->consed_directory->chromat_dir;
}

sub fasta_dir {
    return $_[0]->consed_directory->fasta_dir;
}

#< FASTA >#
my %_fastas_and_amplicon_bioseq_methods = (
    reads => 'get_bioseqs_for_raw_reads',
    processed => 'get_bioseqs_for_processed_reads',
    assembly => 'get_assembly_bioseq',
    oriented => 'get_oriented_bioseq',
);

sub amplicon_fasta_types {
    return keys %_fastas_and_amplicon_bioseq_methods;
}

sub amplicon_bioseq_method_for_type {
    return $_fastas_and_amplicon_bioseq_methods{$_[1]};
}

sub fasta_file_for_type {
    my ($self, $type) = @_;

    return sprintf(
        '%s/%s%s.fasta',
        $self->fasta_dir,
        ( defined $self->subject_name ? $self->subject_name.'.' : '' ),
        $type,
    );
}

sub qual_file_for_type {
    return $_[0]->fasta_file_for_type($_[1]).'.qual';
}

#< Amplicons >#
sub get_amplicons {
    my $self = shift;

    my $method = sprintf(
        '_get_amplicons_and_read_names_for_%s_%s', 
        $self->sequencing_center,
        $self->sequencing_platform,
    );
    my $amplicons = $self->$method;
    unless ( $amplicons and %$amplicons ) {
        $self->error_message(
            sprintf('No amplicons found in chromat_dir of directory (%s)', $self->directory) 
        );
        return;
    }

    my @amplicons;
    my $edit_dir = $self->edit_dir;
    for my $name ( keys %$amplicons ) {
        push @amplicons, Genome::AmpliconAssembly::Amplicon->create(
            name => $name,
            reads => $amplicons->{$name},
            directory => $edit_dir,
        );
    }

    return \@amplicons;
}

sub _get_amplicons_and_read_names_for_gsc_sanger {
    my $self = shift;

    my $dh = Genome::Utility::FileSystem->open_directory( $self->chromat_dir )
        or return;

    my %amplicons;
    while ( my $read_name = $dh->read ) {
        next if $read_name =~ m#^\.#;
        $read_name =~ s#\.gz##;
        my $amplicon_name = $self->_get_amplicon_name_for_gsc_sanger_read_name($read_name)
            or next;
        push @{$amplicons{$amplicon_name}}, $read_name;
    }
    $dh->close;

    return \%amplicons;
}

sub _get_amplicons_and_read_names_for_broad_sanger {
    my $self = shift;

    my $dh = Genome::Utility::FileSystem->open_directory( $self->chromat_dir )
        or return;

    my %amplicons;
    while ( my $read_name = $dh->read ) {
        next if $read_name =~ m#^\.#;
        $read_name =~ s#\.gz$##;
        my $amplicon = $self->_get_amplicon_name_for_broad_sanger_read_name($read_name)
            or return;
        push @{$amplicons{$amplicon}}, $read_name;
    }
    
    return  \%amplicons;
}

#< Amplicon Reads >#
sub get_method_for_get_amplicon_name_for_read_name {
    my $self = shift;

    return sprintf(
        '_get_amplicon_name_for_%s_%s_read_name', 
        $self->sequencing_center,
        $self->sequencing_platform,
    );
}

sub _get_amplicon_name_for_gsc_sanger_read_name {
    my ($self, $read_name) = @_;
    
    $read_name =~ /^(.+)\.[bg]\d+$/
        or return;

    return $1;
}

sub _get_amplicon_name_for_broad_sanger_read_name {
    my ($self, $read_name) = @_;

    $read_name =~ s#\.T\d+$##;
    $read_name =~ s#[FR](\w\d\d?)$#\_$1#; # or next;

    return $read_name;
}

sub get_all_amplicons_reads_for_read_name {
    my ($self, $read_name) = @_;

    unless ( $read_name ) {
        $self->error_message("No read name given to get all reads for amplicon.");
        return;
    }

    my $amp_method = $self->get_method_for_get_amplicon_name_for_read_name;
    my $amplicon_name = $self->$amp_method($read_name);
    unless ( $amplicon_name ) {
        $self->error_message("Can't get amplicon name for read name ($read_name)");
        return;
    }

    my $reads_method = $self->get_method_for_get_all_amplicons_reads_for_read_name;
    my @read_names = $self->$reads_method($amplicon_name);
    unless ( @read_names ) {
        $self->error_message("No reads found for amplicon name ($amplicon_name)");
        return;
    }

    
    return @read_names;
}

sub get_method_for_get_all_amplicons_reads_for_read_name {
    sprintf(
        '_get_all_reads_for_%s_%s_amplicon',
        $_[0]->sequencing_center,
        $_[0]->sequencing_platform,
    );
}

sub _get_all_reads_for_gsc_sanger_amplicon {
    my ($self, $amplicon_name) = @_;
    
    my $chromat_dir = $self->chromat_dir;
    my @read_names;
    for my $read_name ( glob("$chromat_dir/$amplicon_name.*") ) {
        $read_name =~ s#$chromat_dir/##;
        $read_name =~ s#\.gz##;
        push @read_names, $read_name;
    }

    return @read_names;
}

sub _get_all_reads_for_broad_sanger_amplicon {
    die "Not implemented\n";
}

#< Contamination Screening >#
sub contamination_dir {
    return $_[0]->directory.'/contamination';
}

sub contamination_reads_dir {
    return $_[0]->contamination_dir.'/reads';
}
    
sub amplicon_fasta_file_for_contamination_screening {
    return $_[0]->contamination_dir.'/amplicon_reads.fasta';
}

sub create_contamination_dir_structure {
    my $self = shift;

    my $contamination_dir = $self->contamination_dir;
    unless ( -d $contamination_dir ) {
        return unless Genome::Utility::FileSystem->create_directory($contamination_dir);
    }

    my $reads_dir = $self->contamination_reads_dir;
    unless ( -d $reads_dir ) {
        return unless Genome::Utility::FileSystem->create_directory($reads_dir);
    }

    return 1;
}

sub create_amplicon_fasta_file_for_contamination_screening {
    my $self = shift;

    my $amplicons = $self->get_amplicons
        or return;

    $self->create_contamination_dir_structure
        or return;
    
    my $fasta_file = $self->amplicon_fasta_file_for_contamination_screening;
    unlink $fasta_file if -e $fasta_file;
    my $fasta_writer = Bio::SeqIO->new(
        '-file' => '>'.$fasta_file,
        '-fomat' => 'fasta',
    );
    for my $amplicon ( @$amplicons ) {
        for my $bioseq ( $amplicon->get_bioseqs_for_processed_reads ) {
            next unless $bioseq->length >= 11;
            $bioseq->seq( uc $bioseq->seq );
            $fasta_writer->write_seq($bioseq);
        }
    }

    return $fasta_file;
}

sub remove_contaminated_amplicons_by_reads_in_file {
    my ($self, $file) = @_;

    $self->create_contamination_dir_structure
        or return;

    my $fh = Genome::Utility::FileSystem->open_file_for_reading($file)
        or return;

    my %amplicons_seen;
    my $chromat_dir = $self->chromat_dir;
    my $contamination_reads_dir = $self->contamination_reads_dir;
    my $amp_method = $self->get_method_for_get_amplicon_name_for_read_name;
    my $reads_method = $self->get_method_for_get_all_amplicons_reads_for_read_name;

    while ( my $read_name = $fh->getline ) {
        chomp $read_name;
        my $amplicon_name = $self->$amp_method($read_name);
        unless ( $amplicon_name ) {
            $self->error_message("Can't get amplicon name for read name ($read_name)");
            return;
        }
        my @read_names = $self->$reads_method($amplicon_name);
        unless ( @read_names ) {
            $self->error_message("Can't get reads for amplicon name ($amplicon_name)");
            return;
        }
        # Move chromats - this effectively removes the amplicon
        # TODO create the amplicon interface here, to support sanger and 454
        #  maybe move this logic to amplicon?
        for my $read_name ( @read_names ) {
            my $from = "$chromat_dir/$read_name.gz"; 
            my $to = "$contamination_reads_dir/$read_name.gz";
            unless ( -e $from ) { # ok, i guess
                $self->error_message("Can't find trace $read_name");
                next;
            }
            unless ( File::Copy::move($from, $to) ) { # not ok
                $self->error_message("Can't move $from to $to\: $!");
                return;
            }
        }
        # TODO move phd_dir edit_dir files?
    }

    return 1;
}

1;

=pod

=head1 Name

Genome::AmpliconAssembly

=head1 Synopsis

=head1 Usage

=head1 Methods

=head2 valid_sequencing_centers

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 default_sequencing_center

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 consed_directory

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_directory_structure

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 edit_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 phd_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 chromat_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 fasta_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_fasta_types

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_bioseq_method_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 fasta_file_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 qual_file_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 get_amplicons

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 contamination_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 contamination_reads_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_fasta_file_for_contamination_screening

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_contamination_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_amplicon_fasta_files_for_contamination_screening

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 read_is_contaminated

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head1 See Also

=head1 Disclaimer

Copyright (C) 2005 - 2009 Genome Center at Washington University in St. Louis

This module is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY or the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

=head1 Author(s)

B<Eddie Belter> I<ebelter@genome.wustl.edu>

=cut

#$HeadURL$
#$Id$

