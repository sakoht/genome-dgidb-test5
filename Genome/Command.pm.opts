$Genome::Command::OPTS_SPEC = [
          '>project',
          \'Genome::Project::Command',
          '>taxon',
          \'Genome::Taxon::Command',
          '>population-group',
          \'Genome::PopulationGroup::Command',
          '>individual',
          \'Genome::Individual::Command',
          '>sample',
          \'Genome::Sample::Command',
          '>library',
          \'Genome::Library::Command',
          '>instrument-data',
          \'Genome::InstrumentData::Command',
          '>processing-profile',
          \'Genome::ProcessingProfile::Command',
          '>model',
          \'Genome::Model::Command',
          '>tools',
          \'Genome::Model::Tools',
          '>disk',
          \'Genome::Disk::Command',
          'help!',
          undef
        ];
