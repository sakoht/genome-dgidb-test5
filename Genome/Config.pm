package Genome::Config;

use strict;
use warnings;

sub user_email {
    my $self = shift;
    my $user = shift;
    $user ||= $ENV{USER};
    return $user . '@genome.wustl.edu';
}

# operating directories

sub root_directory {
    $ENV{GENOME_MODEL_ROOT} || '/gscmnt/839/info/medseq';
}

sub data_directory {
     $ENV{GENOME_MODEL_DATA} || '/gscmnt/sata363/info/medseq';
}

# links

sub model_links_directory {
    return shift->root_directory . '/model_links';
}

sub alignment_links_directory {
    return shift->root_directory . '/alignment_links';
}

sub model_comparison_link_directory {
    return shift->root_directory . '/model_comparison_links';
}

# data

sub model_data_directory {
    my $self = shift;
    if (defined($ENV{'GENOME_MODEL_TESTDIR'}) &&
        -e $ENV{'GENOME_MODEL_TESTDIR'}
    ) {
            return $ENV{'GENOME_MODEL_TESTDIR'};
    } else {
            return $self->data_directory .'/model_data';
    }   
}

sub alignment_data_directory {
    return shift->data_directory . '/alignment_data';
}

sub model_comparison_data_directory {
    return shift->data_directory . '/model_comparison_data';
}

# reflection of the different types of models, and their related processing profiles and builds

sub type_names {
    return 
        map { s/\-/ /g; $_ }
        map { Command->_command_name_for_class_word($_) }
        map { s/^Genome\::Model:://; $_ } 
        shift->model_subclass_names;
}

my $use_model_subclasses = 0;
sub _use_model_subclasses {
    # We follow a naming convention which allows us to dynamically list all sub-classes of model.
    # There is some flexibility loss by enforcing the naming convention, but the benefit is reflection.
    # A different config could make a different choice if necessary...
    
    unless ($use_model_subclasses) {
        require Genome::Model;
        my $path = $INC{'Genome/Model.pm'};
        unless ($path) {
            die "failed to find the path for Genome/Model.pm in %INC???";
        }
        $path =~ s/.pm\s*$// or die "no pm on $path?";
        unless (-d $path) {
            die "$path is not a directory?";
        }
        my @possible_subclass_modules = glob("$path/*.pm");
        for my $possible_module (@possible_subclass_modules) {
            my $class = $possible_module;
            $class =~ s/.pm$//;
            $class =~ s/\//\:\:/g;
            $class =~ s/^.*(Genome::Model::[^\:]+)/$1/;
            eval "use $class";
            die "Error using module $class ($possible_module): $@" if $@;
            unless ($class->isa("Genome::Model")) {
                next;
            }
            my $suffix = $class;
            $suffix =~ s/^Genome\::Model:://;
            #$model_subclass_names, $class;
        }
        $use_model_subclasses = 1;
    }
    return 1;
}


=pod

=head1 NAME

Genome::Config - environmental configuration for the genome modeling tools

=head1 DESCRIPTION

This module currently just contains global, hard-coded paths.

For portability, it should use an instance of Genome::Config loadable from an environment variable.

=head1 METHODS

=head2 root_directory 

This is the directory under which all other data is symlinked.
It can be changed with the GENOME_MODEL_ROOT environment variable.

This value is typically constant for one site.  It changes only for testing.

=head2 data_directory 

This is the directory under which new data is actually placed.
It can be changed with the GENOME_MODEL_DATA environment variable.

This value changes over time, as disk fills, and new space is made for
new data.

=head2 model_links_directory

All models directories are given a symlink under this directory.  The symlink 
uses the model's "name".

By default the $root_directory/model_links.

=head2 model_data_directory

The default directory into which new models are placed.  Builds go into 
sub-directories here unless otherwise specified.

By default the $data_directory/model_data.

=head2 alignment_links_directory

All alignment linked under this directory.

By default the $root_directory/alignment_links.

=head2 alignment_data_directory

New alignment data is stored or under this directory.

By default the $data_directory/alignment_data.

=head2 comparison_links_directory

Cross-model comparisons are linked here.

By default $current_links_directory/model_comparisons.

=head2 comparison_data_directory

New cross-model comparisons are stored here.

By default $data_directory/model_comparisons.

=cut

1;

