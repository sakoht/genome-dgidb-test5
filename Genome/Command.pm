package Genome::Command;

use strict;
use warnings;

use Genome;

use Data::Dumper;
require File::Basename;

class Genome::Command {
    is => 'Command',
    english_name => 'genome command',
};

our @SUB_COMMAND_DATA = (
    #'project'               => 'Genome::Project::Command',
    #'sample'                => 'Genome::Sample::Command',
    #'population-group'      => 'Genome::PopulationGroup::Command',
    'instrument-data'       => 'Genome::InstrumentData::Command',
    'processing-profile'    => 'Genome::ProcessingProfile::Command',
    'model'                 => 'Genome::Model::Command',
    'tools'                 => 'Genome::Model::Tools',
);
our %SUB_COMMAND_CLASSES = @SUB_COMMAND_DATA;

for my $class ( values %SUB_COMMAND_CLASSES ) {
    eval("use $class;");
    die $@ if $@; 
}

#< Command Naming >#
sub command_name {
    my $class = ref($_[0]) || $_[0];
    return $class->SUPER::command_name unless $class eq __PACKAGE__;
    return 'genome';
}

sub command_name_brief {
    my $class = ref($_[0]) || $_[0];
    return $class->SUPER::command_name_brief unless $class eq __PACKAGE__;
    return 'genome';
}

#< Sub Command Stuff >#
sub is_sub_command_delegator {
    return 1;
}

sub sub_command_classes {
    return values %SUB_COMMAND_CLASSES;
}

sub class_for_sub_command {
    my $class = ref($_[0]) || $_[0];
    return $class->SUPER::class_for_sub_command unless $class eq __PACKAGE__;
    return $SUB_COMMAND_CLASSES{$_[1]};
}

1;

#$HeadURL$
#$Id$

