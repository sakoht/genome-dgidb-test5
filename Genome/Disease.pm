package Genome::Disease;

use strict;
use warnings;

use Genome;
class Genome::Disease {
    type_name => 'disease',
    table_name => 'DISEASE',
    id_by => [
        id => { is => 'NUMBER', len => 20 },
    ],
    has => [
        common_name            => { is => 'VARCHAR2', len => 20, is_optional => 1 },
        disease_parent_disease => { is => 'Genome::Disease', id_by => 'parent_disease_id', constraint_name => 'DISEASE_PARENT_FK' },
        full_name              => { is => 'VARCHAR2', len => 256 },
        parent_disease_id      => { is => 'NUMBER', len => 20, is_optional => 1 },
    ],
    schema_name => 'GMSchema',
    data_source => 'Genome::DataSource::GMSchema',
    doc => 'One row per disease which is an explicit subject of study.',
};

1;
