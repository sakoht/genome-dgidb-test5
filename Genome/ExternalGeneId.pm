package Genome::ExternalGeneId;

use strict;
use warnings;

use Genome;

class Genome::ExternalGeneId {
    type_name => 'genome external gene id',
    table_name => 'EXTERNAL_GENE_ID',
    id_by => [
        egi_id => { is => 'NUMBER' },
    ],
    has => [
        gene_id => { is => 'NUMBER' },
        id_type => { is => 'String' },
        id_value => { is => 'String' },
        build => {
                    is => "Genome::Model::Build",
                    id_by => 'build_id',
        },
        gene => {
            calculate_from => [qw/ gene_id build_id/],
            calculate => q|
                Genome::GeneId->get(gene_id => $gene_id, build_id => $build_id);
            |,
        },
    ],
    schema_name => 'files',
    data_source => 'Genome::DataSource::ExternalGeneIds',
};

1;

