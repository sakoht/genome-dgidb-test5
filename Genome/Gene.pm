package Genome::Gene;

use strict;
use warnings;

use Genome;

class Genome::Gene {
    type_name => 'genome gene',
    table_name => 'GENE',
    id_by => [
        gene_id => { 
            is => 'Text' 
        },
        species => { is => 'varchar',
            is_optional => 1,
        },
        source => { is => 'VARCHAR',
            is_optional => 1,
        },
        version => { is => 'VARCHAR',
            is_optional => 1,
        },
    ],
    has => [
        hugo_gene_name => { 
            is => 'Text',
            is_optional => 1,
        },
        strand => {
            is => 'Text',
            valid_values => ['+1', '-1', 'UNDEF'],
        },
        data_directory => {
            is => "Path",
        },
    ],
    has_many => [
        transcripts => { 
            calculate_from => [qw/ id data_directory/],
            calculate => q|
                Genome::Transcript->get(gene_id => $id,  data_directory => $data_directory);
            |,
        },
        external_ids => { 
            calculate_from => [qw/ id data_directory/],
            calculate => q|
                Genome::ExternalGeneId->get(gene_id => $id, data_directory => $data_directory);
            |,
        },
    ],
    schema_name => 'files',
    data_source => 'Genome::DataSource::Genes',
};

sub name
{
    my ($self, $source) = @_;

    my $name = $self->hugo_gene_name;

    return $name if $name;

    my @egis;
    unless ( $source )
    {
        @egis = $self->external_ids;
    }
    elsif ( $source eq "genbank")
    {
        #$egis = $self->external_ids->search({ id_type => 'entrez' });
        @egis = $self->external_ids(id_type => 'entrez');
    }
    else
    {
        #$egis = $self->external_ids->search({ id_type => $source });
        @egis = $self->external_ids(id_type => $source);
    }

    unless ($egis[0]) {
        return '';
    }

    return $egis[0]->id_value;
}


#- EXPRESSIONS -#
sub expressions_by_intensity
{
    my $self = shift;

    # Sort by decrementing intensity
    my @expressions = sort { $b->expression_intensity <=> $a->expression_intensity }
                           $self->expressions;
    return @expressions;
}


1;

