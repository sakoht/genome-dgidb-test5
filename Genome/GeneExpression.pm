# Review gsanders: This can be removed and has not been used for some time

package Genome::GeneExpression;
#:adukes dump

use strict;
use warnings;

use Genome;

class Genome::GeneExpression {
    type_name => 'genome gene expression',
    table_name => 'GENE_EXPRESSION',
    id_by => [
        expression_id => { is => 'Number' },
    ],
    has => [
        expression_intensity => { is => 'FLoat' },
        dye_type => { is => 'String' },
        probe_sequence => { is => 'CLOB' },
        probe_identifier  => { is => 'String' },
        tech_type => { is => 'String' },
        detection => { is => 'String' },
        build => {
                    is => "Genome::Model::Build",
                    id_by => 'build_id',
        },
    ],
    has_many => [
        gene_expressions => {
             calculate_from => [qw/ expression_id build_id/],
             calculate => q|
                Genome::GeneGeneExpression->get(expression_id => $expression_id, build_id => $build_id);
            |,
        },
        genes => { is => 'Genome::Gene', via => 'gene_expressions', to => 'gene' },
    ],
 
    schema_name => 'files',
    data_source => 'Genome::DataSource::GeneExpressions',
};

1;

