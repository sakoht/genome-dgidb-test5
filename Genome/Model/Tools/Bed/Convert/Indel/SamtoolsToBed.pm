package Genome::Model::Tools::Bed::Convert::Indel::SamtoolsToBed;

use strict;
use warnings;

use Genome;

class Genome::Model::Tools::Bed::Convert::Indel::SamtoolsToBed {
    is => ['Genome::Model::Tools::Bed::Convert::Indel'],
};

sub help_brief {
    "Tools to convert samtools indel format to BED.",
}

sub help_synopsis {
    my $self = shift;
    return <<"EOS"
  gmt bed convert indel samtools-to-bed --source indels_all_sequences --output indels_all_sequences.bed
EOS
}

sub help_detail {                           
    return <<EOS
    This is a small tool to take indel calls in samtools format and convert them to a common BED format (using the first four columns).
EOS
}

sub process_source {
    my $self = shift;
    
    my $input_fh = $self->_input_fh;
    
    for my $line (<$input_fh>) {
        my ($chromosome, $position, $star,
            $_calls, $_consensus_quality, $_ref_quality, $_mapping_quality, $_read_depth,
            $indel_call_1, $indel_call_2, @extra) = split("\t", $line);
        
        next unless $star eq '*'; #samtools indel format includes reference lines as well

        for my $indel ($indel_call_1, $indel_call_2) {
            next if $indel eq '*'; #Indicates only one indel call...and this isn't it!
            
            #position => 1-based position of the start of the indel
            #BED uses 0-based position of and after the event
            
            my ($reference, $variant, $start, $stop);
            
            $start = $position; #samtools reports position before indel so +1 then -1 to switch to 0-based.            
            if(substr($indel,0,1) eq '+') {
                $reference = '*';
                $variant = substr($indel,1);
                $stop = $start + 2; #Two positions are included--the base preceding and the base following the insertion event
            } elsif(substr($indel,0,1) eq '-') {
                $reference = substr($indel,1);
                $variant = '*';
                $stop = $start + length($reference);
            } else {
                $self->error_message('Unexpected indel format encountered: ' . $indel);
                return;
            }
        
            $self->write_bed_line($chromosome, $start, $stop, $reference, $variant);
        }
    }
    
    return 1;
}

1;
