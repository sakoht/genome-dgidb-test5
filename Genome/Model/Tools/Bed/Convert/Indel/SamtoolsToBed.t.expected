1	233507473	233507475	*/A
1	235810391	235810392	A/*
1	236555530	236555532	TG/*
1	239352591	239352593	AA/*
2	19605444	19605446	*/CT
2	24169340	24169342	*/A
2	24476029	24476030	A/*
2	29082560	29082562	*/TG
